aws/protocol/rest/header_test.go

```diff
@@ -0,0 +1,213 @@
+			args:   []interface{}{true},
+			expected: map[string][]string{
+				expectedKeyName: {"true"},
+			},
+		},
+		"set json": {
+			header: http.Header{expectedKeyName: []string{"foobar"}},
+			args:   []interface{}{aws.JSONValue{"jsonKey": "jsonValue"}},
+			expected: map[string][]string{
+				expectedKeyName: {"eyJqc29uS2V5IjoianNvblZhbHVlIn0="},
+			},
+		},
+		"set time": {
+			header: http.Header{expectedKeyName: []string{"foobar"}},
+			args:   []interface{}{time.Unix(0, 0), protocol.ISO8601TimeFormatName},
+			expected: map[string][]string{
+				expectedKeyName: {"1970-01-01T00:00:00Z"},
+			},
+		},
+		"set byte slice": {
+			header: http.Header{expectedKeyName: []string{"foobar"}},
+			args:   []interface{}{[]byte("baz")},
+			expected: map[string][]string{
+				expectedKeyName: {"YmF6"},
+			},
+		},
+		"add string": {
+			header: http.Header{expectedKeyName: []string{"other string"}},
+			args:   []interface{}{"string value"},
+			append: true,
+			expected: map[string][]string{
+				expectedKeyName: {"other string", "string value"},
+			},
+		},
+		"add float64": {
+			header: http.Header{expectedKeyName: []string{"1.61803"}},
+			args:   []interface{}{3.14159},
+			append: true,
+			expected: map[string][]string{
+				expectedKeyName: {"1.61803", "3.14159"},
+			},
+		},
+		"add bool": {
+			header: http.Header{expectedKeyName: []string{"false"}},
+			args:   []interface{}{true},
+			append: true,
+			expected: map[string][]string{
+				expectedKeyName: {"false", "true"},
+			},
+		},
+		"add json": {
+			header: http.Header{expectedKeyName: []string{`eyJzb21lS2V5Ijoic29tZVZhbHVlIn0=`}},
+			args:   []interface{}{aws.JSONValue{"jsonKey": "jsonValue"}},
+			append: true,
+			expected: map[string][]string{
+				expectedKeyName: {"eyJzb21lS2V5Ijoic29tZVZhbHVlIn0=", "eyJqc29uS2V5IjoianNvblZhbHVlIn0="},
+			},
+		},
+		"add time": {
+			header: http.Header{expectedKeyName: []string{"1991-09-17T00:00:00Z"}},
+			args:   []interface{}{time.Unix(0, 0), protocol.ISO8601TimeFormatName},
+			append: true,
+			expected: map[string][]string{
+				expectedKeyName: {"1991-09-17T00:00:00Z", "1970-01-01T00:00:00Z"},
+			},
+		},
+		"add byte slice": {
+			header: http.Header{expectedKeyName: []string{"YmFy"}},
+			args:   []interface{}{[]byte("baz")},
+			append: true,
+			expected: map[string][]string{
+				expectedKeyName: {"YmFy", "YmF6"},
+			},
+		},
+	}
+
+	for name, tt := range cases {
+		t.Run(name, func(t *testing.T) {
+			if tt.header == nil {
+				tt.header = http.Header{}
+			}
+
+			hv := newHeaderValue(tt.header, keyName, tt.append)
+
+			if err := setHeader(hv, tt.args); err != nil {
+				t.Fatalf("expected no error, got %v", err)
+			}
+
+			if e, a := tt.expected, hv.header; !reflect.DeepEqual(e, a) {
+				t.Errorf("expected %v, got %v", e, a)
+			}
+		})
+	}
+}
+
+func TestHeaders(t *testing.T) {
+	const prefix = "x-amzn-meta-"
+	cases := map[string]struct {
+		headers  http.Header
```

Comment:
would be good to have a test case that ensures that adding a `headers` doesn't mutate the map key's casing.

e.g. 

```
map[string]string{
     "foo": "fooValue",
}
```
becomes
```
X-Amzn-Meta-foo: fooValue
```

Would address the S3 Metadata casing issue where the SDK drops the original casing of the header causing unexpected behavior with S3 Metadata for an object.

This probably means the `Encoder` needs to set the `http.Header` map/values  directly and not go through the `Set` and `Add` Methods. https://github.com/aws/aws-sdk-go/issues/445

Reply:
For issue #445 this is actually an issue when retrieving objects and metadata from S3. When Go sends the PutObject request the header would be in the form of: `X-Amzn-Meta-Foo`. S3 normalizes this as `x-amzn-meta-foo` which can be confirmed using the S3 console. But when a GetObject is performed, Go is converting the headers received into canonical mime format, which results in the map key for the returned metadata to become titled. This can be seen in this example:

```go
package main

import (
	"bytes"
	"crypto/rand"
	"flag"
	"fmt"
	"io"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/s3"
	"github.com/davecgh/go-spew/spew"
)

var bucket, region string

func init() {
	flag.StringVar(&bucket, "bucket", "", "bucket name")
	flag.StringVar(&region, "region", "", "region name")
}

func main() {
	sess := session.Must(session.NewSession(&aws.Config{Region: &region}))

	uid := uniqueID()

	fmt.Printf("uid: %v\n", uid)

	svc := s3.New(sess)
	_, err := svc.PutObject(&s3.PutObjectInput{
		Bucket: &bucket,
		Key:    &uid,
		Metadata: map[string]*string{
			"key-one": aws.String("value-one"),
			"key-two": aws.String("value-two"),
		},
		Body: bytes.NewReader(make([]byte, 1024)),
	})
	if err != nil {
		panic(fmt.Errorf("service error: %v", err))
	}

	obj, err := svc.GetObject(&s3.GetObjectInput{
		Bucket: &bucket,
		Key:    &uid,
	})
	if err != nil {
		panic(fmt.Errorf("service error: %v", err))
	}

	spew.Dump(obj.Metadata)
}

func uniqueID() string {
	uuid := make([]byte, 16)
	io.ReadFull(rand.Reader, uuid)
	return fmt.Sprintf("%x", uuid)
}
```

This outputs:
```
uid: de8a11a34eccedcc5fcd4b7ec24d6ae0
(map[string]*string) (len=2) {
 (string) (len=7) "Key-One": (*string)(0xc00018ed60)((len=9) "value-one"),
 (string) (len=7) "Key-Two": (*string)(0xc00018ed70)((len=9) "value-two")
}
```

Also viewable if you capture the HTTPS requests:
<img width="633" alt="Screen Shot 2019-12-11 at 1 26 27 PM" src="https://user-images.githubusercontent.com/549813/70661836-fba6e380-1c19-11ea-99ea-3c72312d322a.png">
<img width="641" alt="Screen Shot 2019-12-11 at 1 26 44 PM" src="https://user-images.githubusercontent.com/549813/70661845-02355b00-1c1a-11ea-8498-b5696afe4c52.png">

Reply:
Should the SDK rely on the behavior of S3? or set the key in the casing the user provided it in? I think this is the key question we should use with regard to S3's Metadata. If another service were to add support for `headers` would it have the same behavior?

Reply:
I've modified the encoder to normalize headers as case-insensitive.